src/cmap/connection.ts

```diff
@@ -1171,26 +1009,130 @@ export class ModernConnection extends TypedEventEmitter<ConnectionEvents> {
       };
     }
 
-    const commandOptions: Document = Object.assign(
-      {
-        numberToSkip: 0,
-        numberToReturn: -1,
-        checkKeys: false,
-        // This value is not overridable
-        secondaryOk: readPreference.secondaryOk()
-      },
-      options
-    );
+    const commandOptions = {
+      numberToSkip: 0,
+      numberToReturn: -1,
+      checkKeys: false,
+      // This value is not overridable
+      secondaryOk: readPreference.secondaryOk(),
+      ...options,
+      readPreference // ensure we pass in ReadPreference instance
+    };
 
     const message = this.supportsOpMsg
-      ? new OpMsgRequest(ns.db, cmd, commandOptions)
-      : new OpQueryRequest(ns.db, cmd, commandOptions);
+      ? new OpMsgRequest(db, cmd, commandOptions)
+      : new OpQueryRequest(db, cmd, commandOptions);
+
+    return message;
+  }
+
+  private async sendCommand(
+    message: WriteProtocolMessageType,
+    options: CommandOptions
+  ): Promise<Document> {
+    const { signal } = this.controller;
 
+    signal.throwIfAborted();
+
+    if (typeof options.socketTimeoutMS === 'number') {
+      this.socket.setTimeout(options.socketTimeoutMS);
+    } else if (this.socketTimeoutMS !== 0) {
+      this.socket.setTimeout(this.socketTimeoutMS);
+    }
+
+    let response;
     try {
-      write(this as any as Connection, message, commandOptions, callback);
-    } catch (err) {
-      callback(err);
+      await writeCommand(this, message, {
+        agreedCompressor: this.description.compressor ?? 'none',
+        zlibCompressionLevel: this.description.zlibCompressionLevel,
+        signal
+      });
+
+      if (options.noResponse) return { ok: 1 };
+
+      response = await read(this, { signal });
+    } finally {
+      if (!signal.aborted) this.controller = new AbortController();
```

Comment:
How come we're instantiating a new controller after every successful command?  I was expecting that `this.controller` would be tied to the lifetime of the connection and that when it did abort, it would indicate that the connection was closed with an error

Reply:
( side note - that pattern looks remarkably similar to scope-based resource management....)

Reply:
This does not appear to change much functionally since we would still be creating a new controller, the difference is it does not assign it back to the Connection instance. I agree we probably need something like this, but I would like to take the time to investigate and design that helper tailored to the CSOT functionality when we're working on that rather than in this PR.

Reply:
[openjsf.org/blog/2022/04/29/using-abortsignal-in-node-js](https://openjsf.org/blog/2022/04/29/using-abortsignal-in-node-js/)

The article I found seems to have the solution we may be looking for. Also documented here: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener. `addEventListener` takes a cleanup signal

Reply:
The added abortable helper ensures we clean up our listeners for write. For reads, see slack, I will discuss in a thread tomorrow.